### PR TITLE
Create a SSL CSR

### PR DESCRIPTION
**Pull request by adiroiban**
Problem
======

We want to create CSR + key from command line but also from Python API

Changes
=======

Created 3 new methods. one to generate the commnad line options, another one to to generate the key/csr and another one to take care of writing to disk.

As a drive by I have updated the test runner to print speed reports and wrote a test for self signed certificate so that we have 100% coverage for ssl module.

How to test
========

reviewers: @alibotean 

check that changes make sense

Use the demo tool to generate a csr

### CODE REVIEW

File: chevah/keycert/tests/test_ssl.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            'organization': None,
+            'organization_unit': None,
+            'locality': None,
+            'state': None,
+            'country': None,
+            }, options)
+
+    def test_value(self):
+        """
+        Options are parsed form command line.
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: from

File: chevah/keycert/tests/test_ssl.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+        self.assertStartsWith('usage: test-command key-gen [-h]', error)
+        self.assertEndsWith(
+            '\ntest-command key-gen: '
+            'error: argument --common-name is required\n',
+            error)
+
+    def test_default(self):
+        """
+        It can be initiaized with only a subparser and sub-command name.
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: initialized

File: chevah/keycert/tests/test_ssl.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            '--key-type=rsa',
+            '--key-size=12',
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('key size too small', context.exception.message)
+
+    def test_bad_county_long(self):
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: country

File: chevah/keycert/tests/test_ssl.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+            '--common-name=domain.com',
+            '--country=USA',
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('string too long', context.exception.message)
+
+    def test_bad_county_short(self):
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: country

File: chevah/keycert/tests/test_ssl.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('string too short', context.exception.message)
+
+    def test_default_gen(self):
+        """
+        By default it will serialized the key without password and generate
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: serialize

File: chevah/keycert/tests/test_ssl.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+            crypto.FILETYPE_PEM, result['csr'])
+        self.assertEqual(csr, result['csr_pem'])
+        subject = result['csr'].get_subject()
+        self.assertEqual(u'domain.com', subject.commonName)
+        self.assertIsNone(subject.emailAddress)
+        self.assertIsNone(subject.organizationName)
+        self.assertIsNone(subject.organizationalUnitName)
+        self.assertIsNone(subject.localityName)
+        self.assertIsNone(subject.stateOrProvinceName)
+        self.assertIsNone(subject.countryName)
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: alternative name?

File: chevah/keycert/tests/test_ssl.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+    def test_gen_unicode(self):
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: unicode -> encoding?

File: chevah/keycert/tests/test_ssl.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+            u'name@xn--domain-uro-x77e.com', subject.emailAddress)
+        self.assertEqual(u'OU Nam\u20acuro', subject.organizationName)
+        self.assertEqual(u'OU Unit\u20acuro', subject.organizationalUnitName)
+        self.assertEqual(u'Som\u20acwhere', subject.localityName)
+        self.assertEqual(u'Stat\u20ac', subject.stateOrProvinceName)
+        self.assertEqual(u'GB', subject.countryName)
+
+    def test_encrypted_key(self):
+        """
+        When asked it serialized the key with a password.
```
Comment (alibotean) [inline chevah/keycert/tests/test_ssl.py]: it will serialize

**Comment by alibotean** [comment]
Just some minor comments.

changes-approved

File: chevah/keycert/tests/test_ssl.py (hunk 6 above)
Comment (adiroiban) [inline chevah/keycert/tests/test_ssl.py]: alternative name is installed as an extension and the current API does not allow getting an extension... I know is strange... pyopenssl 0.15 will add this option.. but is not yet released

File: chevah/keycert/tests/test_ssl.py (hunk 7 above)
Comment (adiroiban) [inline chevah/keycert/tests/test_ssl.py]: the test checks how the code behaves when presented with Unicode data.

**Pull request closed without merging**